// necoengine/login/control.js

import {view} from "./view.js" 
import {model} from "./model.js" 

"use strict"

const control = {
  func: {
    login: {
      exec: function(){
        model.login.exec()
      },
      add: function(){
        const loginButton = view.elements.loginButton
        loginButton.addEventListener("click",(e)=>{
          e.preventDefault()
          this.exec()
        },false)
      },
    },
    create: {
      exec: function(){
        model.create.exec()
      },
      add: function(){
        const createButton = view.elements.createButton
        createButton.addEventListener("click",(e)=>{
          e.preventDefault()
          this.exec()
        },false)
      }
    }
  },
  visit: function(){
    model.visit.exec()
  },
  loginPage: function(
     email=document.forms.login.email,
     password = document.forms.login.password,
     button = document.forms.login.button,
     message = document.querySelector("article")
     ){

    model.checkLogined.exec()
    view.setElement("loginEmail", email) 
    view.setElement("loginPassword", password) 
    view.setElement("loginButton", button) 
    view.setElement("loginMessage", message) 
    
    const sessionEmail = window.sessionStorage.getItem("userEmail")
    email.value = sessionEmail || ""

    control.func.login.add()
  },
  createPage: function(
     userName=document.forms.create.username,
     email=document.forms.create.email,
     password = document.forms.create.password,
     confirmPassword = document.forms.create.confirmpassword,
     button = document.forms.create.button,
     message = document.querySelector("article")
     ){

    model.checkLogined.exec()
    view.setElement("createUserName", userName) 
    view.setElement("createEmail", email) 
    view.setElement("createPassword", password) 
    view.setElement("createConfirmPassword", confirmPassword) 
    view.setElement("createButton", button) 
    view.setElement("createMessage", message) 

    control.func.create.add()

  },
  checkLogined:function(callback){
    model.checkLogined.exec(callback)
  },
  setLoginButton: function(pinFlag = true){
    return model.setLoginButton.exec(pinFlag)
 }
}

export {control}
