// necoengine/engine/model.js
"use strict"

const engineGetAppURL = "/necoengine/api/getapp"
const engineSearchURL = "/necoengine/api/search"

export const model = {
  getApp: function(){
    const userEmail = window.localStorage.getItem("userEmail") || "anonymous"
    const url = engineGetAppURL 
    const data = {
      body: JSON.stringify({email:userEmail}),
      cache: "no-cache",
      credentials: "same-origin",
      headers: {"content-type":"application/json"},
      method: "POST",
      mode: "cors",
      redirect: "follow",
      referror: "no-referror",
    }
    const main = async () =>{
      const response = await fetch(url,data)
      const json = await response.json()
      return new Promise((resolve,reject)=>{
        resolve(json)
      })
    }
    return main()
  },
  search:function(res){
    const userEmail = window.localStorage.getItem("userEmail") || "anonymous"
    const searchWords = res.searchWords
    const pageNumber = res.pageNumber
    const body = {
      email:userEmail, 
      searchWords:searchWords, 
      pangeNUmber:pageNumber
    }
    const url = engineSearchURL 
    const data = {
      body: JSON.stringify(body),
      cache: "no-cache",
      credentials: "same-origin",
      headers: {"content-type":"application/json"},
      method: "POST",
      mode: "cors",
      redirect: "follow",
      referror: "no-referror",
    }
    const main = async () =>{
      const response = await fetch(url,data)
      const json = await response.json()
      return new Promise((resolve,reject)=>{
        resolve(json)
      })
    }
    return main()
  }
}
